src/vmm/src/vcpu/mod.rs

```diff
@@ -0,0 +1,264 @@
+        guest_memory
+            .write_obj(boot_pde_addr.raw_value() as u64 | 0x03, boot_pdpte_addr)
+            .map_err(Error::GuestMemory)?;
+
+        // 512 2MB entries together covering VA [0..1GB).
+        // This assumes that the CPU supports 2MB pages (/proc/cpuinfo has 'pse').
+        for i in 0..512 {
+            guest_memory
+                .write_obj((i << 21) + 0x83u64, boot_pde_addr.unchecked_add(i * 8))
+                .map_err(Error::GuestMemory)?;
+        }
+
+        sregs.cr3 = boot_pml4_addr.raw_value() as u64;
+        sregs.cr4 |= X86_CR4_PAE;
+        sregs.cr0 |= X86_CR0_PG;
+
+        self.vcpu_fd.set_sregs(&sregs).map_err(Error::KvmIoctl)
+    }
+
+    /// Configure FPU.
+    pub fn configure_fpu(&self) -> Result<()> {
+        let fpu = kvm_fpu {
+            fcw: 0x37f,
+            mxcsr: 0x1f80,
+            ..Default::default()
+        };
+        self.vcpu_fd.set_fpu(&fpu).map_err(Error::KvmIoctl)
+    }
+
+    /// Configures LAPICs. LAPIC0 is set for external interrupts, LAPIC1 is set for NMI.
+    pub fn configure_lapic(&self) -> Result<()> {
+        let mut klapic = self.vcpu_fd.get_lapic().map_err(Error::KvmIoctl)?;
+
+        let lvt_lint0 = get_klapic_reg(&klapic, APIC_LVT0);
+        set_klapic_reg(
+            &mut klapic,
+            APIC_LVT0,
+            set_apic_delivery_mode(lvt_lint0, APIC_MODE_EXTINT),
+        );
+        let lvt_lint1 = get_klapic_reg(&klapic, APIC_LVT1);
+        set_klapic_reg(
+            &mut klapic,
+            APIC_LVT1,
+            set_apic_delivery_mode(lvt_lint1, APIC_MODE_NMI),
+        );
+
+        self.vcpu_fd.set_lapic(&klapic).map_err(Error::KvmIoctl)
+    }
+
+    /// vCPU emulation loop.
+    #[allow(clippy::if_same_then_else)]
+    pub fn run(&mut self) {
+        match self.vcpu_fd.run() {
+            Ok(exit_reason) => {
+                match exit_reason {
+                    VcpuExit::Shutdown | VcpuExit::Hlt => {
+                        println!("Guest shutdown: {:?}. Bye!", exit_reason);
+                        if stdin().lock().set_canon_mode().is_err() {
+                            eprintln!("Failed to set canon mode. Stdin will not echo.");
+                        }
+                        unsafe { libc::exit(0) };
+                    }
+                    VcpuExit::IoOut(addr, data) => {
+                        if 0x3f8 <= addr && addr < (0x3f8 + 8) {
+                            // Write at the serial port.
+                            if self
+                                .serial
+                                .lock()
+                                .unwrap()
+                                .0
+                                .write((addr - 0x3f8) as u8, data[0])
+                                .is_err()
+                            {
+                                eprintln!("Failed to write to serial port");
+                            }
+                        } else if 0x060 <= addr && addr < (0x060 + 5) {
```

Comment:
Aren't only 0x60, 0x61 and 0x64 - i8042 ports?

Reply:
I actually have no idea :smiley: thanks for pointing it out! Is there a recommended link I could link here as "ground truth"?

Reply:
This one seems trustworthy: https://web.archive.org/web/20200602023256/https://wiki.osdev.org/%228042%22_PS/2_Controller#PS.2F2_Controller_IO_Ports, but I couldn't find some "official" list of the IO ports meaning.